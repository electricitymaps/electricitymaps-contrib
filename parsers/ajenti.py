"""
Initial PR https://github.com/electricitymaps/electricitymaps-contrib/pull/2456
Discussion thread https://github.com/electricitymaps/electricitymaps-contrib/issues/636
A promotion webpage for King's Island energy production is here: https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/king-island
As of 09/2020, it embeds with <iframe> the URI https://data.ajenti.com.au/KIREIP/index.html

About the data, the feed we get seems to be counters with a 2 seconds interval.
That means that if we fetch these counters every 15 minutes, we only are reading "instantaneous" metters that could differ from the total quantity of energies at play.
To get the very exact data, we would need to have a parser running constanty to collect those 2-sec interval counters.
"""

import json
from datetime import datetime
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

from requests import Session
from signalr import Connection

ZONE_PARAMS = {
    # King Island
    # https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/king-island
    "AU-TAS-KI": {
        "hub": "TagHub",
        "method": "Dashboard",
        "tz": "Australia/Currie",
        "source": "hydro.com.au",  # Iframe: https://data.ajenti.com.au/KIREIP/index.html
    },
    # Flinders Island
    # https://github.com/electricitymaps/electricitymaps-contrib/issues/2533
    # https://en.wikipedia.org/wiki/Flinders_Island
    # https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/flinders-island
    "AU-TAS-FI": {
        "hub": "flindershub",
        "method": "SendDashboard",
        "tz": "Australia/Hobart",
        "source": "hydro.com.au",
    },
    # Rottnest Island
    # https://github.com/electricitymaps/electricitymaps-contrib/issues/2534
    # https://en.wikipedia.org/wiki/Rottnest_Island
    # https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/rottnest-island
    "AU-WA-RI": {
        "hub": "HogsHub",
        "method": "SendDashboard",
        "tz": "Australia/Perth",
        "source": "hydro.com.au",
    },
}


class SignalR:
    def __init__(self, url):
        self.url = url

    def update_res(self, msg):
        if msg != {}:
            self.res = msg

    def get_value(self, hub, method):
        self.res = {}
        with Session() as session:
            # create a connection
            connection = Connection(self.url, session)
            chat = connection.register_hub(hub)
            chat.client.on(method, self.update_res)
            connection.start()
            connection.wait(3)
            connection.close()
            return self.res


def parse_payload(logger: Logger, payload) -> dict:
    technologies_parsed = {
        "biomass": 0,
        "battery": 0,
        "coal": 0,
        "flywheel": 0,
        "gas": 0,
        "hydro": 0,
        "nuclear": 0,
        "oil": 0,
        "solar": 0,
        "wind": 0,
        "geothermal": 0,
        "unknown": 0,
    }
    if "technologies" not in payload:
        raise KeyError(
            f"No 'technologies' in payload\n" f"serie : {json.dumps(payload)}"
        )
    else:
        logger.debug(f"serie : {json.dumps(payload)}")
    for technology in payload["technologies"]:
        # rename upstream key to match our payload needs
        if technology["id"] == "diesel":
            technology["id"] = "oil"

        assert technology["unit"] == "kW"
        # The upstream API gives us kW, we need MW
        technologies_parsed[technology["id"]] = int(technology["value"]) / 1000

    # King's Island uses some percentage of biodiesel; since we account differently traditionnal oil & biomass, here we do math to separate both
    if "biodiesel" in payload:
        # sanity check to be sure that we are not remaking data that the data source could otherwise give us
        assert "biomass" and "oil" != 0
        technologies_parsed["biomass"] = (
            technologies_parsed["oil"] * payload["biodiesel"]["percent"] / 100
        )
        technologies_parsed["oil"] = (
            technologies_parsed["oil"] * (100 - payload["biodiesel"]["percent"]) / 100
        )

    logger.debug(f"production : {json.dumps(technologies_parsed)}")

    return technologies_parsed


# Both keys battery and flywheel are negative when storing energy, and positive when feeding energy to the grid
def format_storage_techs(technologies_parsed):
    # sometimes we don't get those keys. For instance Rottnest island don't have any storage.
    if "battery" and "flywheel" in technologies_parsed:
        storage_techs = technologies_parsed["battery"] + technologies_parsed["flywheel"]
    else:
        storage_techs = 0
    _battery_production = (
        storage_techs if storage_techs > 0 else 0
    )  # TODO: Should this just be removed?
    _battery_storage = storage_techs if storage_techs < 0 else 0


def sum_storage_techs(technologies_parsed):
    storage_techs = technologies_parsed["battery"] + technologies_parsed["flywheel"]

    return storage_techs


def fetch_production(
    zone_key: str = "AU-TAS-KI",
    session: Session | None = None,
    target_datetime=None,
    logger: Logger = getLogger(__name__),
) -> dict:
    if target_datetime is not None:
        raise NotImplementedError(
            "The datasource currently implemented is only real time"
        )

    # get the specific zone parameters
    try:
        hub, dashboard, tz, source = (
            ZONE_PARAMS[zone_key]["hub"],
            ZONE_PARAMS[zone_key]["method"],
            ZONE_PARAMS[zone_key]["tz"],
            ZONE_PARAMS[zone_key]["source"],
        )
    except KeyError as e:
        raise KeyError("The zone " + zone_key + " isn't implemented") from e

    payload = SignalR("https://data.ajenti.com.au/live/signalr").get_value(
        hub, dashboard
    )
    technologies_parsed = parse_payload(logger, payload)
    storage_techs = sum_storage_techs(technologies_parsed)

    return {
        "zoneKey": zone_key,
        "datetime": datetime.now(tz=ZoneInfo(tz)),
        "production": {
            "biomass": technologies_parsed["biomass"],
            "coal": technologies_parsed["coal"],
            "gas": technologies_parsed["gas"],
            "hydro": technologies_parsed["hydro"],
            "nuclear": technologies_parsed["nuclear"],
            "oil": technologies_parsed["oil"],
            "solar": technologies_parsed["solar"],
            "wind": 0
            if technologies_parsed["wind"] < 0 and technologies_parsed["wind"] > -0.1
            else technologies_parsed[
                "wind"
            ],  # If wind between 0 and -0.1 set to 0 to ignore self-consumption
            "geothermal": technologies_parsed["geothermal"],
            "unknown": technologies_parsed["unknown"],
        },
        "storage": {
            "battery": storage_techs
            * -1  # Somewhat counterintuitively,to ElectricityMap positive means charging and negative means discharging
        },
        "source": source,
    }


if __name__ == "__main__":
    print(fetch_production())
