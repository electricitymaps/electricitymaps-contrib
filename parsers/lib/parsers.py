import importlib

from electricitymap.contrib.config import EXCHANGES_CONFIG, ZONES_CONFIG

# Prepare all parsers
CONSUMPTION_PARSERS = {}
PRODUCTION_PARSERS = {}
PRODUCTION_PER_MODE_FORECAST_PARSERS = {}
PRODUCTION_PER_UNIT_PARSERS = {}
EXCHANGE_PARSERS = {}
PRICE_PARSERS = {}
CONSUMPTION_FORECAST_PARSERS = {}
GENERATION_FORECAST_PARSERS = {}
EXCHANGE_FORECAST_PARSERS = {}
PRODUCTION_CAPACITY_PARSERS = {}
DAYAHEAD_LOCATIONAL_MARGINAL_PRICE_PARSERS = {}
REALTIME_LOCATIONAL_MARGINAL_PRICE_PARSERS = {}

PARSER_KEY_TO_DICT = {
    "consumption": CONSUMPTION_PARSERS,
    "production": PRODUCTION_PARSERS,
    "productionPerUnit": PRODUCTION_PER_UNIT_PARSERS,
    "productionPerModeForecast": PRODUCTION_PER_MODE_FORECAST_PARSERS,
    "exchange": EXCHANGE_PARSERS,
    "price": PRICE_PARSERS,
    "dayaheadLocationalMarginalPrice": DAYAHEAD_LOCATIONAL_MARGINAL_PRICE_PARSERS,
    "realtimeLocationalMarginalPrice": REALTIME_LOCATIONAL_MARGINAL_PRICE_PARSERS,
    "consumptionForecast": CONSUMPTION_FORECAST_PARSERS,
    "generationForecast": GENERATION_FORECAST_PARSERS,
    "exchangeForecast": EXCHANGE_FORECAST_PARSERS,
    "productionCapacity": PRODUCTION_CAPACITY_PARSERS,
}


def _parser_key_to_parser_folder(parser_key: str):
    return (
        "electricitymap.contrib.capacity_parsers"
        if parser_key == "productionCapacity"
        else "parsers"
    )


# Read all zones
for zone_id, zone_config in ZONES_CONFIG.items():
    for parser_key, v in zone_config.get("parsers", {}).items():
        mod_name, fun_name = v.split(".")
        mod = importlib.import_module(
            f"{_parser_key_to_parser_folder(parser_key)}.{mod_name}"
        )
        PARSER_KEY_TO_DICT[parser_key][zone_id] = getattr(mod, fun_name)


# Read all exchanges
for exchange_id, exchange_config in EXCHANGES_CONFIG.items():
    for parser_key, v in exchange_config.get("parsers", {}).items():
        mod_name, fun_name = v.split(".")
        mod = importlib.import_module(f"parsers.{mod_name}")
        PARSER_KEY_TO_DICT[parser_key][exchange_id] = getattr(mod, fun_name)
