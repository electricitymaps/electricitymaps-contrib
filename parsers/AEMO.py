import io
import re
import zipfile
from datetime import datetime
from logging import Logger, getLogger
from typing import Any
from urllib.parse import urljoin
from zoneinfo import ZoneInfo

import pandas as pd
from bs4 import BeautifulSoup
from requests import Session

from electricitymap.contrib.lib.models.event_lists import TotalConsumptionList
from electricitymap.contrib.lib.models.events import EventSourceType
from electricitymap.contrib.lib.types import ZoneKey

SOURCE = "aemo.com.au"

ZONE_KEY_TO_REGION = {
    "AU-NSW": "NSW1",
    "AU-QLD": "QLD1",
    "AU-SA": "SA1",
    "AU-TAS": "TAS1",
    "AU-VIC": "VIC1",
    "AU-WA": "WEM",  # This zone is not implemented yet
}

ZONE_KEY_TO_TIMEZONE = {
    "AU-NSW": ZoneInfo("Australia/Sydney"),
    "AU-QLD": ZoneInfo("Australia/Brisbane"),
    "AU-SA": ZoneInfo("Australia/Adelaide"),
    "AU-TAS": ZoneInfo("Australia/Hobart"),
    "AU-VIC": ZoneInfo("Australia/Melbourne"),
    "AU-WA": ZoneInfo("Australia/Perth"),  # This zone is not implemented yet
}

# TODO, what about the other zone in Australia (AU-WA)? Check remaining zone


def find_document(session, target_datetime):
    # Fetch the directory listing
    base_url = "http://nemweb.com.au/Reports/CURRENT/Operational_Demand/FORECAST_HH/"
    response = session.get(base_url)

    # Parse with BeautifulSoup
    soup = BeautifulSoup(response.text, "html.parser")

    # String datetime
    target_date_str = target_datetime.strftime("%Y%m%d")

    # Find matching links
    matching_links = soup.find_all(
        "a",
        href=re.compile(
            rf"PUBLIC_FORECAST_OPERATIONAL_DEMAND_HH_{target_date_str}\d+_\d+\.zip"
        ),
    )

    if matching_links:
        target_link = matching_links[-1]  # Get the last (most recent) link
        full_url = urljoin(base_url, target_link["href"])  # Construct full URL

        file_response = session.get(full_url)

        with zipfile.ZipFile(io.BytesIO(file_response.content)) as z:
            csv_filename = z.namelist()[0]
            with z.open(csv_filename) as f:
                df = pd.read_csv(f)
                return df
    else:
        print("No matching files found")


def fetch_consumption_forecast(
    zone_key: ZoneKey,  # "AU-NSW", "AU-QLD", "AU-SA", "AU-TAS", "AU-VIC", "AU-WA"
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    """Consumption forecast in MW every half an hour for 10 days ahead.
    Only for NSW1, QND1, SA1, TAS1, VIC1 zones."""
    session = session or Session()

    if target_datetime is None:
        target_datetime = datetime.now(tz=ZONE_KEY_TO_TIMEZONE[zone_key])

    df = find_document(session, target_datetime)

    # Transform dataframe
    df.columns = df.iloc[0]
    df = df.iloc[1:-1].reset_index(drop=True)

    #
    region = ZONE_KEY_TO_REGION.get(zone_key)
    all_consumption_events = df[
        df["REGIONID"] == region
    ]  # all events with a datetime and a consumption value
    consumption_list = TotalConsumptionList(logger)
    for _, event in all_consumption_events.iterrows():
        datetime_object = datetime.strptime(
            event["INTERVAL_DATETIME"], "%Y/%m/%d %H:%M:%S"
        ).replace(tzinfo=ZONE_KEY_TO_TIMEZONE[zone_key])

        consumption_list.append(
            zoneKey=zone_key,
            datetime=datetime_object,
            consumption=float(
                event["OPERATIONAL_DEMAND_POE50"]
            ),  # 50% probability of exceedance operational demand forecast value
            source=SOURCE,
            sourceType=EventSourceType.forecasted,
        )
    return consumption_list.to_list()


if __name__ == "__main__":
    """Main method, never used by the electricityMap backend, but handy for testing."""

    print(fetch_consumption_forecast("AU-NSW"))
    print(fetch_consumption_forecast("AU-QLD"))
    print(fetch_consumption_forecast("AU-SA"))
    print(fetch_consumption_forecast("AU-TAS"))
    print(fetch_consumption_forecast("AU-VIC"))
    print(
        fetch_consumption_forecast("AU-WA")
    )  # Not implemented yet. It returns an empty list
