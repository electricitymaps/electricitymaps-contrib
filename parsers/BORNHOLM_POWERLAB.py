#!/usr/bin/env python3
from datetime import datetime
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

from requests import Session

from electricitymap.contrib.lib.models.event_lists import (
    ExchangeList,
    ProductionBreakdownList,
)
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey

PRODUCTION_MAPPING = {
    "wind": "wind_turbines",
    "biomass": "factory",
    "solar": "solar_cells",
}
LATEST_DATA_URL = "https://bornholm.powerlab.dk/visualizer/latestdata"

SOURCE = "bornholm.powerlab.dk"
TIMEZONE = ZoneInfo("UTC")


def fetch_production(
    zone_key: ZoneKey = ZoneKey("DK-BHM"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known production mix (in MW) of a given country."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")
    if session is None:
        session = Session()
    response = session.get(LATEST_DATA_URL).json()
    production = ProductionBreakdownList(logger)
    production.append(
        zoneKey=zone_key,
        datetime=datetime.fromtimestamp(response["latest"], tz=TIMEZONE),
        source=SOURCE,
        production=ProductionMix(
            biomass=response["sub"]["factory"],
            solar=response["sub"]["solar_cells"],
            wind=response["sub"]["wind_turbines"],
        ),
    )
    return production.to_list()


def fetch_exchange(
    zone_key1: ZoneKey = ZoneKey("DK-BHM"),
    zone_key2: ZoneKey = ZoneKey("SE-SE4"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known power exchange (in MW) between two countries."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")
    if session is None:
        session = Session()
    response = session.get(LATEST_DATA_URL).json()
    exchange = ExchangeList(logger)
    sorted_zone_keys = sorted([zone_key1, zone_key2])
    direction = 1 if zone_key1 == sorted_zone_keys[0] else -1
    exchange.append(
        zoneKey=ZoneKey("->".join(sorted_zone_keys)),
        datetime=datetime.fromtimestamp(response["latest"], tz=TIMEZONE),
        source=SOURCE,
        netFlow=direction * response["sub"]["seacable"],
    )

    return exchange.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""

    print("fetch_production() ->")
    print(fetch_production())
    print("fetch_exchange(DK-BHM, SE-SE4) ->")
    print(fetch_exchange())
