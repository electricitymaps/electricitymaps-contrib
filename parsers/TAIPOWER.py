#!/usr/bin/env python3
from datetime import datetime, timedelta
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

import pandas as pd
from requests import Session

from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
from electricitymap.contrib.lib.models.events import ProductionMix, StorageMix
from parsers.lib.config import refetch_frequency
from parsers.lib.exceptions import ParserException

SOURCE = "taipower.com.tw"
TIMEZONE = ZoneInfo("Asia/Taipei")
PRODUCTION_URL = (
    "http://www.taipower.com.tw/d006/loadGraph/loadGraph/data/genary_eng.json"
)


@refetch_frequency(timedelta(days=1))
def fetch_production(
    zone_key: str = "TW",
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")

    response = session.get(PRODUCTION_URL)
    if not response.status_code == 200:
        raise ParserException(
            "TW",
            f"Query failed with status code {response.status_code} and {response.content}",
        )

    data = response.json()

    dt = data[""]
    prodData = data["dataset"]

    dt = datetime.strptime(dt, "%Y-%m-%d %H:%M").replace(tzinfo=TIMEZONE)

    objData = pd.DataFrame(prodData)

    columns = [
        "fueltype",
        "additional_1",
        "name",
        "capacity",
        "output",
        "percentage",
        "additional_2",
        "additional_3",
    ]
    assert len(objData.iloc[0]) == len(columns), "number of input columns changed"
    objData.columns = columns
    objData["fueltype"] = objData.fueltype.str.split("<b>").str[1]
    objData["fueltype"] = objData.fueltype.str.split("</b>").str[0]
    objData.loc[:, ["capacity", "output"]] = objData[["capacity", "output"]].apply(
        pd.to_numeric, errors="coerce"
    )
    if objData["fueltype"].str.contains("OTHERRENEWABLEENERGY").any():
        if objData["name"].str.contains("Geothermal").any():
            objData.loc[objData["name"].str.contains("Geothermal"), "fueltype"] = (
                "Geothermal"
            )
        if objData["name"].str.contains("Biofuel").any():
            objData.loc[objData["name"].str.contains("Biofuel"), "fueltype"] = "Biofuel"

    assert not objData.capacity.isna().all(), (
        "capacity data is entirely NaN - input column order may have changed"
    )
    assert not objData.output.isna().all(), (
        "output data is entirely NaN - input column order may have changed"
    )

    objData.drop(
        columns=["additional_1", "name", "additional_2", "percentage", "additional_3"],
        axis=1,
        inplace=True,
    )
    # summing because items in returned object are for each power plant and operational units
    production = pd.DataFrame(objData.groupby("fueltype").sum())
    production.columns = ["capacity", "output"]

    # check output values coincide with total capacity by fuel type
    check_values = production.output <= production.capacity
    modes_with_capacity_exceeded = production[~check_values].index.tolist()
    for mode in modes_with_capacity_exceeded:
        logger.warning(f"Capacity exceeded for {mode} in {zone_key} at {dt}")

    # For storage, note that load will be negative, and generation positive.
    # We require the opposite
    PRODUCTION_MODE_MAPPING = {
        "biomass": ["Biofuel"],
        "coal": ["COAL", "IPPCOAL"],
        "gas": ["LNG", "IPPLNG"],
        "geothermal": ["Geothermal"],
        "oil": ["OIL", "DIESEL"],
        "hydro": ["HYDRO"],
        "nuclear": ["NUCLEAR"],
        "solar": ["SOLAR"],
        "wind": ["WIND"],
        "unknown": ["COGEN"],
    }
    STORAGE_MODE_MAPPING = {
        "hydro": ["PUMPINGLOAD", "PUMPINGGEN"],
    }

    production_breakdown = ProductionBreakdownList(logger)
    production_mix = ProductionMix()
    for mode, parser_modes in PRODUCTION_MODE_MAPPING.items():
        parser_modes_in_df = [
            parser_mode
            for parser_mode in parser_modes
            if parser_mode in production.index
        ]
        production_mix.add_value(mode, production.loc[parser_modes_in_df].output.sum())
    storage_mix = StorageMix()
    for mode, storage_modes in STORAGE_MODE_MAPPING.items():
        storage_modes_in_df = [
            storage_mode
            for storage_mode in storage_modes
            if storage_mode in production.index
        ]
        storage_mix.add_value(
            mode, -1 * production.loc[storage_modes_in_df].output.sum()
        )
    production_breakdown.append(
        zone_key,
        dt,
        SOURCE,
        production_mix,
        storage_mix,
    )

    capacity = {}
    for mode, parser_modes in PRODUCTION_MODE_MAPPING.items():
        parser_modes_in_df = [
            parser_mode
            for parser_mode in parser_modes
            if parser_mode in production.index
        ]
        capacity[mode] = production.loc[parser_modes_in_df].capacity.sum()

    return [{**e, **{"capacity": capacity}} for e in production_breakdown.to_list()]


if __name__ == "__main__":
    print(fetch_production())
