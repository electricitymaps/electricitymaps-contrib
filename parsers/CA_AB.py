#!/usr/bin/env python3

"""Parse the Alberta Electric System Operator's (AESO's) Energy Trading System
(ETS) website.
"""

# Standard library imports
import csv
import re
import urllib.parse
from datetime import datetime, timedelta
from logging import Logger, getLogger
from typing import Any
from zoneinfo import ZoneInfo

# Third-party library imports
from requests import Session

from electricitymap.contrib.lib.models.event_lists import ExchangeList, PriceList
from electricitymap.contrib.lib.types import ZoneKey

# Local library imports
from parsers.lib import validation

DEFAULT_ZONE_KEY = ZoneKey("CA-AB")
MINIMUM_PRODUCTION_THRESHOLD = 10  # MW
TIMEZONE = ZoneInfo("Canada/Mountain")
URL = urllib.parse.urlsplit("http://ets.aeso.ca/ets_web/ip/Market/Reports")
URL_STRING = urllib.parse.urlunsplit(URL)


def fetch_exchange(
    zone_key1: str = DEFAULT_ZONE_KEY,
    zone_key2: str = "CA-BC",
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    """Request the last known power exchange (in MW) between two countries."""
    if target_datetime:
        raise NotImplementedError("Currently unable to scrape historical data")
    session = session or Session()
    response = session.get(
        f"{URL_STRING}/CSDReportServlet", params={"contentType": "csv"}
    )
    interchange = dict(csv.reader(response.text.split("\r\n\r\n")[4].splitlines()))
    flows = {
        f"{DEFAULT_ZONE_KEY}->CA-BC": interchange["British Columbia"],
        f"{DEFAULT_ZONE_KEY}->CA-SK": interchange["Saskatchewan"],
        f"{DEFAULT_ZONE_KEY}->US-MT": interchange["Montana"],
        f"{DEFAULT_ZONE_KEY}->US-NW-NWMT": interchange["Montana"],
    }
    sorted_zone_keys = ZoneKey("->".join(sorted((zone_key1, zone_key2))))
    if sorted_zone_keys not in flows:
        raise NotImplementedError(f"Pair '{sorted_zone_keys}' not implemented")
    exchanges = ExchangeList(logger)
    exchanges.append(
        zoneKey=sorted_zone_keys,
        datetime=get_csd_report_timestamp(response.text),
        netFlow=float(flows[sorted_zone_keys]),
        source=URL.netloc,
    )
    return exchanges.to_list()


def fetch_price(
    zone_key: ZoneKey = DEFAULT_ZONE_KEY,
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    """Request the last known power price of a given country."""
    if target_datetime:
        raise NotImplementedError("Currently unable to scrape historical data")
    session = session or Session()
    response = session.get(
        f"{URL_STRING}/SMPriceReportServlet", params={"contentType": "csv"}
    )
    prices = PriceList(logger)
    for row in csv.reader(response.text.split("\r\n\r\n")[2].splitlines()[1:]):
        if row[1] != "-":
            date, hour = row[0].split()
            prices.append(
                zoneKey=zone_key,
                datetime=datetime.strptime(
                    f"{date} {int(hour) - 1}", "%m/%d/%Y %H"
                ).replace(tzinfo=TIMEZONE)
                + timedelta(hours=1),
                price=float(row[1]),
                source=URL.netloc,
                currency="CAD",
            )
    return prices.to_list()


def fetch_production(
    zone_key: str = DEFAULT_ZONE_KEY,
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> dict[str, Any] | None:
    """Request the last known production mix (in MW) of a given country."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")
    session = session or Session()
    response = session.get(
        f"{URL_STRING}/CSDReportServlet", params={"contentType": "csv"}
    )
    generation = {
        row[0]: {
            "MC": float(row[1]),  # maximum capability
            "TNG": float(row[2]),  # total net generation
        }
        for row in csv.reader(response.text.split("\r\n\r\n")[3].splitlines())
    }
    return validation.validate(
        {
            "capacity": {
                "gas": generation["COGENERATION"]["MC"]
                + generation["COMBINED CYCLE"]["MC"]
                + generation["GAS FIRED STEAM"]["MC"]
                + generation["SIMPLE CYCLE"]["MC"],
                "wind": generation["WIND"]["MC"],
                "solar": generation["SOLAR"]["MC"],
                "hydro": generation["HYDRO"]["MC"],
                "biomass": generation["OTHER"]["MC"],
                "battery storage": generation["ENERGY STORAGE"]["MC"],
            },
            "datetime": get_csd_report_timestamp(response.text),
            "production": {
                "gas": generation["COGENERATION"]["TNG"]
                + generation["COMBINED CYCLE"]["TNG"]
                + generation["GAS FIRED STEAM"]["TNG"]
                + generation["SIMPLE CYCLE"]["TNG"],
                "wind": generation["WIND"]["TNG"],
                "solar": generation["SOLAR"]["TNG"],
                "hydro": generation["HYDRO"]["TNG"],
                "biomass": generation["OTHER"]["TNG"],
            },
            "source": URL.netloc,
            "storage": {
                "battery": generation["ENERGY STORAGE"]["TNG"],
            },
            "zoneKey": zone_key,
        },
        logger,
        floor=MINIMUM_PRODUCTION_THRESHOLD,
        remove_negative=True,
    )


def get_csd_report_timestamp(report):
    """Get the timestamp from a current supply/demand (CSD) report."""
    return datetime.strptime(
        re.search(r'"Last Update : (.*)"', report).group(1), "%b %d, %Y %H:%M"
    ).replace(tzinfo=TIMEZONE)


if __name__ == "__main__":
    # Never used by the electricityMap backend, but handy for testing.
    print("fetch_production() ->")
    print(fetch_production())
    print("fetch_price() ->")
    print(fetch_price())
    print(f"fetch_exchange({DEFAULT_ZONE_KEY}, CA-BC) ->")
    print(fetch_exchange(DEFAULT_ZONE_KEY, "CA-BC"))
    print(f"fetch_exchange({DEFAULT_ZONE_KEY}, CA-SK) ->")
    print(fetch_exchange(DEFAULT_ZONE_KEY, "CA-SK"))
    print(f"fetch_exchange({DEFAULT_ZONE_KEY}, US-MT) ->")
    print(fetch_exchange(DEFAULT_ZONE_KEY, "US-MT"))
    print(f"fetch_exchange({DEFAULT_ZONE_KEY}, US-NW-NWMT) ->")
    print(fetch_exchange(DEFAULT_ZONE_KEY, "US-NW-NWMT"))
