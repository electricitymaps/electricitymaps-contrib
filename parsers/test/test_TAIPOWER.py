from requests_mock import GET

from electricitymap.contrib.lib.types import ZoneKey
from parsers.TAIPOWER import PRODUCTION_URL, fetch_production


def test_production(adapter, session, snapshot):
    with open("parsers/test/mocks/TAIPOWER/genary.json", "rb") as mock_file:
        adapter.register_uri(
            GET,
            PRODUCTION_URL,
            content=mock_file.read(),
        )

    assert snapshot == fetch_production(
        zone_key=ZoneKey("TW"),
        session=session,
    )
