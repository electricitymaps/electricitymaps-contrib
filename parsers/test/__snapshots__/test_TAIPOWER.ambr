# serializer version: 1
# name: test_production
  list([
    dict({
      'capacity': dict({
        'biomass': 38.2,
        'coal': 14197.2,
        'gas': 18060.6,
        'geothermal': 7.2,
        'hydro': 2101.9999999999995,
        'nuclear': 1902.0,
        'oil': 1592.5,
        'solar': 11304.099999999999,
        'unknown': 626.9,
        'wind': 1032.4,
      }),
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2023, 12, 29, 19, 10, tzinfo=zoneinfo.ZoneInfo(key='Asia/Taipei')),
      'production': dict({
        'biomass': 25.8,
        'coal': 6367.6,
        'gas': 14010.6,
        'geothermal': 2.6,
        'hydro': 458.7,
        'nuclear': 1894.4,
        'oil': 314.5,
        'solar': 0.0,
        'unknown': 1494.6,
        'wind': 1712.7,
      }),
      'source': 'taipower.com.tw',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
        'hydro': -935.7,
      }),
      'zoneKey': 'TW',
    }),
  ])
# ---
